lsp-completion.el

```diff
@@ -495,19 +495,20 @@ Others: TRIGGER-CHARS"
 
     (cond
      ((equal lsp-completion-provider :none))
-     ((and (null lsp-completion-provider)
-           (functionp 'company-lsp))
-      (progn
-        (company-mode 1)
-        (add-to-list 'company-backends 'company-lsp)
-        (setq-local company-backends (remove 'company-capf company-backends))))
+
+     ((and lsp-completion-provider
+           (fboundp 'company-mode)
+           (member 'company-lsp company-backends))
+      (message "Warning: `company-lsp` is not supported anymore, using `company-capf` as the `lsp-completion-provider`.")
+      (company-mode 1)
+      (add-to-list 'company-backends 'company-capf))
```

Comment:
This will not likely work for default `company-backends` and with `clangd`.
Here is the default `company-backends`
```
(company-bbdb company-eclim company-semantic company-clang company-xcode company-cmake company-capf company-files
              (company-dabbrev-code company-gtags company-etags company-keywords)
              company-oddmuse company-dabbrev)
```

Since `company-clang` is before `company-capf`, the `add-to-list` will still keep the position of `company-capf` the same after `company-clang`. Thus `company-clang` will be used for `cpp` file instead.
Moreover, `add-to-list` modify the global `company-backends`, which is affecting user configuration for something not in `lsp-mode`.

Maybe you should just do `(setq company-backends '(company-capf))`

Reply:
Ok, but this can remove any other `company-backend` right like `company-yasnippet`?

Reply:
We have a separate issue for that, and most likely it will be fixed in company-mode. Lets get that in faster because ATM company-lsp is broken.

Reply:
At least I think we should make `company-backends` become local variable here instead of global `add-to-list`. But that's just nit pick, so feel free to go ahead.

Reply:
> Ok, but this can remove any other `company-backend` right like `company-yasnippet`?

Also, unless `company-yasnippet` is merged into `company-capf` as multi backends, it will be something like either snippet or  `capf`, which prevents the other backend from displayed.

Reply:
> At least I think we should make `company-backends` become local variable here instead of global `add-to-list`. But that's just nit pick, so feel free to go ahead.

Why are you insisting on breaking users' configs?  We are running this for several months with no complaints. Furthermore, there were a number of complaints regarding the decision to use buffer-local for company-backends and company-lsp.

Reply:
> Why are you insisting on breaking users' configs? We are running this for several months with no complaints. Furthermore, there were a number of complaints regarding the decision to use buffer-local for company-backends and company-lsp.

Which is the complain about making `company-backends` buffer local?

I tried to search for `company-backends local` in this repo and found no related issue so far. Whereas I did find a complain about `company-backends` is changed. https://github.com/emacs-lsp/lsp-mode/issues/1209. That issue can be regarded as should just disable auto configure.

Let me point out that setting `(add-to-list 'company-backends 'company-capf)` is mostly wasteful, since by default `company-capf` is included in `company-backends`, but if user ever remove `company-capf` (I don't think anyone does this, thus no complain so far), then the `add-to-list` will modify the global `company-backends` and affect other files out of `lsp-mode`.
Purely from logic POV, we've tried to set local variable for most of configuration, why it's not good to make `company-backends` local variable here?

Reply:
> 
> I tried to search for `company-backends local` in this repo and found no related issue so far. Whereas I did find a complain about `company-backends` is changed. #1209. That issue can be regarded as should just disable auto configure.

Here: https://github.com/emacs-lsp/lsp-mode/commit/7c55ba55af5c5e061285993c952074cbc52bfa67#commitcomment-31759080 . There were also in reddit, gitter, spacemacs channel, etc. So, when I implemented the initial capf support I decided not to create a local variable. Since then, there wasn't a single complaint in lsp-mode repo for that. 

> Purely from logic POV, we've tried to set local variable for most of configuration, why it's not good to make `company-backends` local variable here?

because it doesn't bring any value for that particular case and it will break users configs for no reason.

Reply:
I think you misunderstand my point.
Here is my point out:

> Let me point out that setting `(add-to-list 'company-backends 'company-capf)` is mostly wasteful, since by default `company-capf` is included in `company-backends`, but if user ever remove `company-capf` (I don't think anyone does this, thus no complain so far), then the `add-to-list` will modify the global `company-backends` and affect other files out of `lsp-mode`.

And the example you're giving is about setting `company-backends` to `company-lsp` only, and user is complain (maybe just asking why).
Here in this issue https://github.com/emacs-lsp/lsp-mode/issues/1209, it actually complain about **adding** into `company-backends`.

Just to make it clear, I said it would be nice to just set `(setq-local company-backends '(company-capf))`, and it should set `(make-local-variable 'company-backends)`.

Let me point out another thing again, currently `(add-to-list 'company-backends 'company-capf)` is no different from `(setq 'company-backends 'company-capf)` if user has no stand-alone `company-capf` in `company-backends`.
The reason why is because `lsp-completion-at-point-function` will try to complete everything and leave no later backend to run.

Reply:
Here is another example, if your `company-backends` is configured as `(company-nxml company-capf)`, and you visit and `xml` file with autoconfigure, you will get completions from `company-nxml` instead of `capf`.

Reply:
> Here is another example, if your `company-backends` is configured as `(company-nxml company-capf)`, and you visit and `xml` file with autoconfigure, you will get completions from `company-nxml` instead of `capf`.

Indeed  - in this case, the user prefers company-nxml and most likely uses company-capf with M-x company-other-backend. What you are proposing will break user's config for no reason. I am locking the conversation.

Reply:
**Ok, while I'm hoping you can reconsider, I respect your choice and direction to go forward.**

The example I gave out before can be just user configures `company-mode` long time before to work with XML file, and now discover `lsp-mode`, however, the completion is not working OOTB. Same thing for `company-clang`.

Please also consider this as well
> Let me point out another thing again, currently (add-to-list 'company-backends 'company-capf) is no different from (setq 'company-backends 'company-capf) if user has no stand-alone company-capf in company-backends.
The reason why is because lsp-completion-at-point-function will try to complete everything and leave no later backend to run.

Well, again, thanks for your patient and opinions.